plugins/org.eclipse.n4js.transpiler/src/org/eclipse/n4js/n4jsx/transpiler/utils/JSXBackendHelper.java

```diff
@@ -168,74 +139,33 @@ public String jsxBackendModuleQualifiedName(Resource resource) {
 	 * @throws RuntimeException
 	 *             when no JSX backend is available
 	 */
-	private final String getAnyBackend() {
+	private String getAnyBackend() {
 		return jsxBackends.keySet().stream().findAny()
 				.orElseThrow(() -> new RuntimeException("Compiler cannot locate JSX backend to use for this resource"));
 	}
 
 	/**
 	 * Populates {@link #jsxBackends} with backends visible from provided resource.
 	 */
-	private final void populateBackendsCache(Resource resource) {
-		jsxBackends.putAll(
-				visibleBackends(resource, JSXBackendHelper::looksLikeReactUri)
-						.stream()
-						.collect(Collectors.toMap(
-								uri -> qualifiedNameConverter.toString(nameComputer.getQualifiedModuleName(uri)),
-								Function.identity(),
-								// IDE-2505
-								JSXBackendHelper::stubMerger)));
-
-	}
-
-	/**
-	 * Collects all {@link IContainer}s visible from provided resources. Returned collection is filter with provided
-	 * predicate.
-	 *
-	 * Similar to {code DefaultGlobalScopeProvider.getVisibleContainers(Resource)}.
-	 *
-	 * @param resource
-	 *            for which we look for visible containers
-	 * @param predicate
-	 *            used to filter collected containers
-	 * @return filtered set of visible containers
-	 */
-	private Set<URI> visibleBackends(Resource resource, Predicate<URI> predicate) {
-		Set<URI> backends = new HashSet<>();
-		List<IContainer> visibleContainers = xtextUtil.getVisibleContainers(resource);
-		visibleContainers.stream()
-				.map(container -> container.getResourceDescriptions())
-				.forEach(resourceDscriptions -> resourceDscriptions.iterator()
-						.forEachRemaining(candidateResourceDescription -> {
-							URI uri = candidateResourceDescription.getURI();
-							if (predicate.test(uri)) {
-								backends.add(uri);
-							}
-						}));
-		return Collections.unmodifiableSet(backends);
+	private void populateBackendsCache(Resource resource) {
+		jsxBackends.clear();
+		addReactToBackendsCache(resource);
 	}
 
-	/**
-	 * Helper function that allows to deal with duplicate resources with that have the same FQN in the same scope. In
-	 * general this is configuration error and, normally, should not happen.Due to lack of proper validations, it can
-	 * happen. Validations preventing this situation are expected to be added with IDE-2505. Once that is done this
-	 * merger should be removed.
-	 *
-	 * For reasons described above, we provide just some merging function, without much thought about its internals.
-	 *
-	 */
-	private static URI stubMerger(URI first, URI second) {
-		if (first.lastSegment().compareToIgnoreCase(second.lastSegment()) > 0)
-			return first;
-		return second;
+	private void addReactToBackendsCache(Resource resource) {
+		TModule reactModule = reactHelper.lookUpReactTModule(resource);
+		if (reactModule != null) {
+			URI reactModuleURI = reactModule.eResource().getURI();
+			jsxBackends.put(ReactHelper.REACT_PROJECT_ID, reactModuleURI);
```

Comment:
If we delegate to react helper, which is caching lookup result, do we need to cache here?

Reply:
The caching in JSXBackendHelper is problematic, anyway, and probably has to be removed; but this has to be investigated further as part of #358.

But I'll add a comment there to make sure it won't be forgotten!